### ISSUE TITLE
MINIO_BROWSER=off but sensitive data still visible with anonymous bucket policy

### ISSUE THREAD

**Issue by narrington** [minio/minio#8402]
We need to share a large volume of files with multiple customers. Our customers cant be bothered with logins, and the files need to remain available for a year or more. So the files need to be shared via direct links from public buckets...

So we set the bucket policy to public... and customers are able to see the bucket contents by deleting the filename from the link. also they are able to UPLOAD files... not good.
So we set the bucket policy to download... but they can STILL see the bucket contents.
So we set the environment variable "MINIO_BROWSER=off".,. the customers are STILL able to see the content of the buckets, only now it's listed in XML...
So we set the bucket policy to none... now the links come back "access denied" still in XML, but at least the bucket itself isn't accessible.

what do I have to do to secure these buckets so the bucket contents remain invisible, while allowing individual files to be accessed?

for someone who's about to ask, there's no real danger to people seeing the files, or the buckets, it just looks bad. we'd PREFER an error 404 when a customer attempts to modify the link we send them, but an access denied xml error is sufficient.

**Comment by harshavardhana** [comment]
what is your bucket policy here @narrington do you not want List access to your customers?

**Comment by narrington** [comment]
@harshavardhana, we're just using the canned policies (none, download, upload, public). I'm willing to set a custom policy, but I'm having trouble understanding the guide... I'm a bit of a script kiddie: not very good with actual coding.

**Comment by kannappanr** [comment]
@narrington You can set a custom policy like the following for a bucket named `my-bucket`

```
{
 "Statement": [
  {
   "Action": [
    "s3:GetObject"
   ],
   "Effect": "Allow",
   "Principal": {
    "AWS": [
     "*"
    ]
   },
   "Resource": [
    "arn:aws:s3:::my-bucket/*"
   ]
  }
 ],
 "Version": "2012-10-17"
}

```
This policy will give public access to the objects within the bucket names `my-bucket`. There is also a way to share the objects using `pre-signed` urls. Please look at https://docs.min.io/docs/minio-client-complete-guide.html and look at `mc share` Though the maximum time that a shared URL to an object is valid only for 7 days. 

It will be easier for us to answer if you can reach out to us at https://slack.min.io for such questions

**Comment by kannappanr** [comment]
@narrington Closing this issue as answered. Please feel free to reach out to us at https://slack.min.io for further questions

**Comment by narrington** [comment]
Note to anyone who cares, yes. the custom policy posted by @kannappanr worked like a champ. 

I created a set of text files ~/.minio/policies/bucketname.json and used his code as a template for each. then applied the policies with: 
`mc policy set none my-minio/my-bucket1`
`mc policy set-json ~/.minio/policies/my-bucket1.json my-minio/my-bucket1`
`mc policy set none my-minio/my-bucket2`
`mc policy set-json ~/.minio/policies/my-bucket2.json my-minio/my-bucket2`
`mc policy set none my-minio/my-bucket3`
`mc policy set-json ~/.minio/policies/my-bucket3.json my-minio/my-bucket3`

Then I removed "MINIO_BROWSER=off".

Now customers don't get an XML file list from the buckets, and mangled URLs drop them to a login prompt for minio (which is MUCH prettier than an XML error)...

I'd STILL like a proper 404 message for bad URLs, but that's a feature request for later!

Thanks @kannappanr!
